text/0000-async-project-goal-2024h2.md

```diff
@@ -0,0 +1,275 @@
+    // rust-lang/rust#123432).
+}
+```
+
+The goal for this year to be able to 
+
+* support some "async equivalent" to `Fn`, `FnMut`, and `FnOnce` bounds
+    * this should be usable in all the usual places
+* support some way to author async closure expressions
+
+These features should be sufficient to support methods like `filter` above.
+
+The details (syntax, precise semantics) will be determined via experimentation and subject to RFC.
+
+### Stabilize trait for async iteration
+
+There has been extensive discussion about the best form of the trait for async iteration (sometimes called `Stream`, sometimes `AsyncIter`, and now being called `AsyncGen`). We believe the design space has been sufficiently explored that it should be possible to author an RFC laying out the options and proposing a specific plan.
+
+### Resolve the ["send bound"][sb] problem
+
+Although async functions in traits were stabilized, there is currently no way to write a generic function that requires only impls where the returned futures are `Send`. This blocks the use of async function in traits in some core ecosystem crates, such as [tower](https://crates.io/crates/tower), which want to work across all kinds of async executors. This problem is called the ["send bound"][sb] problem and there has been extensive discussion of the various ways to solve it. [RFC #3654][] has been opened proposing one solution and describing why that path is preferred. Our goal for the year is to adopt *some* solution on stable.
+
+[RFC #3654]: https://github.com/rust-lang/rfcs/pull/3654
+
+### Author draft RFC for async vision
+
+We plan to revise the [Async Vision Doc][AVD] and restructure it as a draft RFC, most likely to be approved by the [Lang][] and [Libs-API][] teams (we do not necessarily expect that RFC to be accepted by end of year). Our observation is that the previous version of the async vision doc, which was never RFC'd, never attained the legitimacy of being the "plan of record". In addition, a number of things have changed in the intervening years (for example, async functions in traits are now stable) and we are in a position to identify clearer next steps. The 
+
+[AVD]: https://rust-lang.github.io/wg-async/vision.html
+
+This RFC will lay out a "plan of attack" for async, including both obvious good things (similar to [async closures][]) but also "known unknowns" and ways to resolve them. Areas the RFC is expected to cover are as follows:
+
+[Making Async Rust Reliable]: https://tmandry.gitlab.io/blog/posts/making-async-reliable/
+
+* Status quo, covering biggest challenges
+    * Lack of strong learning material
+    * Common idioms contain footguns that cause unexpected failures (see e.g., Tyler's blog post [Making Async Rust Reliable][])
+    * Low-level performance hurdles, such as large future sizes and downsides of the poll model
+    * Fragmentation between runtimes
+* Design axioms to pursue for async (see e.g. axioms proposed)
+* Goals, some variant of
```

Comment:
How do you think these goals should relate to the "shiny future" and axioms proposed in this goal RFC? Are the goals based on those, but slightly more specific, or with concrete and measurable outcomes?

Reply:
Good question. I imagined these goals as being roughly equivalent to the shiny-future shown here, I think. It's possible we should remove some of this outline, but I did want to use it to advertise and drive alignment early ont the expected "rough direction".

Reply:
I decided to just "redirect" to the shiny future